r2-core/src/main/java/com/linkedin/r2/message/RequestContext.java

```diff
@@ -43,7 +50,7 @@ public RequestContext()
     // this class publicly exposes the _localAttrs through getLocalAttrs() and
     // returned Map that must support null values because there is plenty of code
     // that is using this property.
-    _localAttrs = Collections.synchronizedMap(new HashMap<>());
```

Comment:
How about we just switch to `ConcurrentHashMap`. From the javadoc
>Similarly, Iterators, Spliterators and Enumerations return elements reflecting the state of the hash table at some point at or since the creation of the iterator/enumeration. They do not throw [ConcurrentModificationException](https://docs.oracle.com/javase/8/docs/api/java/util/ConcurrentModificationException.html). However, iterators are designed to be used by only one thread at a time.

Reply:
My concerns with `ConccurentHashMap` are first that it will introduce more overhead into a very low fundamental part of the R2 stack, and second that it allows for a copy to be created that is not consistent with a single point in time. Since the iteration behavior is "return each value at most once", the copy could end up with two values that never coexisted at the same time in the original. This is a change in the contract the current behavior provides, and seeing as how we see quite a few of these errors in EXC tickets a lot of people are relying on concurrent edits to the attribute map and could be potentially impacted by a change in behavior.

Reply:
Have you verified that ConcurrentHashMap has a higher overhead? I'd be somewhat surprised if it did, as reads are not blocking.

I'm not sure if the concern over the consistency is an issue. Currently if anyone is making changes to the map during an iteration the exception is thrown, so there is no consistent behavior during modifications. Using a concurrent map won't guarantee consistency, but we also don't know what the impact of allowing that will be.

Reply:
Alternatively, if the main pain point is ```When making a RequestContext copy we must iterate over the collection``` how about just adding a thread-safe copy() method?

Reply:
Just to share what I got from looking at the stack traces. Looks like the common pattern is like the following:

```streamRequest(request, new RequestContext(requestContext)```

Then the copy constructor of RequestContext will need to copy the localAttrs syncMap, and if at the same time the original context is modified in another thread for whatever reason, the copy will hit the exception. @bbarkley 

And to make it worse, we offers an unsafe API of "getLocalAttrs()" which directly returns that internal map to users, and that would make it impossible(?) for us to fix the concurrency locally inside RequestContext .

Reply:
@jpstewart from the doc https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedMap(java.util.Map) , seems like we can at least fix the pattern above by fixing the copy constructor? We can add a synchronized loop there to copy the map.

Reply:
It looks like you are correct, I tested and verified this works. I expected it would not since the `SynchronizedMap` synchronizes on a mutex object held privately, but actually `Collections.synchronizedMap(m)` will use m itself as the mutex, so this works.